BIP153: SENDTEMPLATE (#1937)
bitcoin/bips: bip-0153.md

```
  BIP: 153
  Layer: Peer Services
  Title: Sharing Block Templates with Peers
  Author: Anthony Towns <aj@erisian.com.au>
  Comments-Summary: No comments yet.
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0153
  Status: Draft
  Type: Standards Track
  Created: 2025-09-05
  License: BSD-3-Clause
  Post-History: https://gnusha.org/pi/bitcoindev/aJvZwR_bPeT4LaH6@erisian.com.au/T/#u

>                 https://delvingbitcoin.org/t/sharing-block-templates/1906

Reviewer: Missing header: `Requires: 141, 152`

any payload data received should be ignored.
3. If a node implements this BIP, it must send the `sendtemplate`
message after `version` and before `verack`.

### `gettemplate`

1. The `gettemplate` message is defined as a message with `pchCommand ==
"gettemplate"`.
2. No payload data should be provided.
3. If the node received a `sendtemplate` message, it may send a
`gettemplate` message at any time after `verack`, in order
to request a new template.
3. A `gettemplate` message must not be sent if a `sendtemplate` message
was not already received.
4. Upon receipt of a `gettemplate` message, a node may reply with a
`template` message.

### `template`

1. The `template` message is defined as a message with `pchCommand ==
"template"` and a payload of a serialized `HeaderAndShortIDs` message
(see [BIP-152][BIP-152]) containing a recent template.
2. A `template` message may only be sent in response to a `gettemplate`
message.
3. The `header` hash included in the payload must be a unique hash,
and should be the hash of a block made up of those transactions.
4. The transactions making up the payload of the template should be

> valid for block inclusion -- for example all unconfirmed parents of a

Reviewer: ```suggestion
valid for block inclusion -- for example, all unconfirmed parents of a
```

transaction should be included before the transaction, and consensus
limits should be respected.
5. The cumulative weight of the transactions making up the
payload must not exceed eight million weight units as defined in
[BIP-141][BIP-141]. Note that this is twice the block weight limit, and
does not include the overhead that assembling those transactions into
a block would involve (header, coinbase and tx count serialization),
so this can be slightly more than two blocks' worth of transactions.
6. After sending a `template` message, a node should accept a

> `sendblocktxns` message that specify the template's `header` hash, and

Reviewer: ```suggestion
`sendblocktxns` message that specifies the template's `header` hash, and
```

respond with a `blocktxns` message including the appropriate transaction
data, exactly as if it had sent a `cmpctblock` message with the same
payload as the `template` message. It should be able to provide such
transactions for a moderate amount of time after sending a template

> (eg, five minutes). If it cannot provide a correct `blocktxns` response,

Reviewer: ```suggestion
(e.g., five minutes). If it cannot provide a correct `blocktxns` response,
```

Author: Ditto for ie

worth of transactions. This is chosen so that even if two blocks are
mined in quick succession, prior templates that were shared using this
protocol are likely to have included the transactions from both blocks.

If blocks are found randomly, it is expected that 5% of blocks will
be found within 30 seconds of the previous block, which is about 7 per
day; so this seems frequent enough to address explicitly. In contrast,
the chances of finding two blocks within 30 seconds of a block is only
0.12%, so even in ideal circumstances, is only expected to occur about
once per week, and thus further optimisation does not seem warranted.

### Forward compatibility

It is likely that improvements can be made to this design, particularly
to further reduce the bandwidth needed for updating templates. This
could be achieved by:

  * defining a new payload for the `template` message with a new, more
    efficient encoding of (changes to) the template data
  * adding a payload to the `gettemplate` message so that peers can
    indicate they will accept a new encoding of the template
  * adding a payload to the `sendtemplate` message to indicate that the
    `gettemplate` payload will be accepted

> ## Recommendations

Reviewer 2: Perhaps this could use some recommendations re privacy implications? 

IIRC, there was some push back augments devs re [BIP 35 (`mempool` message)](https://github.com/bitcoin/bips/blob/master/bip-0035.mediawiki), as it can be used to implement relay correlation and tagging attacks against peers. As the message enables a peer to obtain the entire mempool of another peer, it can feasibly be used by an adversary to figure out: who relayed what, which transactions may have originated directly from the node, etc. 

Some related PRs from the bitcoind repo:
  * https://github.com/bitcoin/bitcoin/pull/27426 -- an attempt to remove the command all together, but was later abandoned. 
  * https://github.com/bitcoin/bitcoin/pull/7840 -- restrcited the type of peers that bitcoind will respond to w/ responses to the `mempool` message (white list applied), the start of some of the inv trickle/batch re address privacy concerns 

Also pool operators may want to filter out transactions that they received via out of band transaction submission.

Reviewer 3: > As the message enables a peer to obtain the entire mempool of another peer, it can feasibly be used by an adversary to figure out: who relayed what, which transactions may have originated directly from the node, etc.

Are you here talking about the `mempool` message or this proposal? It seems to me that the top two blocks of the mempool would be significantly more difficult/expensive to exploit for fingerprinting than the whole mempool, especially if nodes repeatedly use this new peer service to synchronize their top mempool.

```
  BIP: 153
  Layer: Peer Services
  Title: Sharing Block Templates with Peers
  Author: Anthony Towns <aj@erisian.com.au>
  Comments-Summary: No comments yet.
  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-0153
  Status: Draft
  Type: Standards Track
  Created: 2025-09-05
  License: BSD-3-Clause
  Post-History: https://gnusha.org/pi/bitcoindev/aJvZwR_bPeT4LaH6@erisian.com.au/T/#u
                https://delvingbitcoin.org/t/sharing-block-templates/1906

>   Requires: 141, 152

Reviewer 3: ```suggestion
  BIP: 153
  Layer: Peer Services
  Title: Sharing Block Templates with Peers
  Authors: Anthony Towns <aj@erisian.com.au>
  Status: Draft
  Type: Specification
  Assigned: 2025-09-05
  License: BSD-3-Clause
  Discussion: https://gnusha.org/pi/bitcoindev/aJvZwR_bPeT4LaH6@erisian.com.au/T/#u
               https://delvingbitcoin.org/t/sharing-block-templates/1906
  Requires: 141, 152
```
